File: server/storage.ts
import { randomUUID } from "crypto";
import type {
  Conversation,
  Message,
  Endpoint,
  Domain,
  Site,
  Config,
  InsertDomain,
  InsertEndpoint,
} from "@shared/schema";

export interface IStorage {
  getConversations(): Promise<Conversation[]>;
  getConversation(id: string): Promise<Conversation | undefined>;
  createConversation(endpointId: string, title: string, domainId?: string, siteId?: string): Promise<Conversation>;
  addMessage(conversationId: string, message: Omit<Message, "id">): Promise<Message>;
  updateConversation(id: string, updates: Partial<Conversation>): Promise<Conversation | undefined>;
  deleteConversation(id: string): Promise<boolean>;
  
  getDomains(): Promise<Domain[]>;
  getDomain(id: string): Promise<Domain | undefined>;
  createDomain(domain: InsertDomain): Promise<Domain>;
  updateDomain(id: string, updates: Partial<InsertDomain>): Promise<Domain | undefined>;
  deleteDomain(id: string): Promise<boolean>;
  
  getSites(): Promise<Site[]>;
  getSite(id: string): Promise<Site | undefined>;
  
  getEndpoints(domainId?: string): Promise<Endpoint[]>;
  getEndpoint(id: string): Promise<Endpoint | undefined>;
  createEndpoint(endpoint: InsertEndpoint): Promise<Endpoint>;
  updateEndpoint(id: string, updates: Partial<InsertEndpoint>): Promise<Endpoint | undefined>;
  deleteEndpoint(id: string): Promise<boolean>;
  
  getConfig(): Promise<Config>;
  setConfig(config: Config): Promise<Config>;
}

export class MemStorage implements IStorage {
  private conversations: Map<string, Conversation>;
  private endpoints: Map<string, Endpoint>;
  private domains: Map<string, Domain>;
  private sites: Map<string, Site>;
  private config: Config;

  constructor() {
    this.conversations = new Map();
    this.endpoints = new Map();
    this.domains = new Map();
    this.sites = new Map();
    this.config = {
      systemPrompt: "You are a helpful AI assistant for Anglo American. Provide clear, accurate, and helpful responses based on the conversation context.",
    };

    this.initializeDefaultDomains();
    this.initializeDefaultSites();
    this.initializeDefaultEndpoints();
  }

  private initializeDefaultDomains() {
    const defaultDomains: Domain[] = [
      {
        id: "generic",
        name: "General Assistant",
        description: "General purpose AI assistant for any queries",
        systemPrompt: "You are a helpful AI assistant for Anglo American. Provide clear, accurate, and helpful responses.",
        icon: "Bot",
      },
      {
        id: "mining-ops",
        name: "Mining Operations",
        description: "Mining operations, production, and safety analytics",
        systemPrompt: "You are an expert AI assistant specializing in mining operations for Anglo American. Help with production metrics, equipment performance, shift planning, safety protocols, and operational efficiency. Provide data-driven insights for mining operations.",
        icon: "HardHat",
      },
      {
        id: "geological",
        name: "Geological Services",
        description: "Geological surveys, ore body analysis, and exploration",
        systemPrompt: "You are a geological expert AI for Anglo American. Assist with geological survey analysis, ore body modeling, mineral exploration data, grade control, and geological mapping. Provide scientific insights based on geological data.",
        icon: "Mountain",
      },
      {
        id: "processing",
        name: "Mineral Processing",
        description: "Ore processing, metallurgy, and plant operations",
        systemPrompt: "You are a mineral processing specialist AI for Anglo American. Help with ore processing optimization, metallurgical analysis, plant throughput, recovery rates, and processing efficiency. Focus on improving extraction and processing outcomes.",
        icon: "Factory",
      },
      {
        id: "sustainability",
        name: "Sustainability & ESG",
        description: "Environmental, social, and governance reporting",
        systemPrompt: "You are a sustainability expert AI for Anglo American. Assist with ESG reporting, environmental impact assessments, carbon footprint analysis, water management, community relations, and sustainability metrics. Help drive responsible mining practices.",
        icon: "Leaf",
      },
      {
        id: "supply-chain",
        name: "Supply Chain",
        description: "Logistics, procurement, and supply chain analytics",
        systemPrompt: "You are a supply chain specialist AI for Anglo American. Help with logistics optimization, procurement analytics, inventory management, vendor performance, and supply chain efficiency. Provide insights to streamline operations.",
        icon: "Truck",
      },
      {
        id: "finance",
        name: "Finance & Analytics",
        description: "Financial analysis, budgeting, and cost optimization",
        systemPrompt: "You are a financial analyst AI for Anglo American. Assist with financial reporting, cost analysis, budget forecasting, capital allocation, and financial performance metrics. Help optimize financial outcomes for mining operations.",
        icon: "TrendingUp",
      },
    ];

    defaultDomains.forEach((domain) => {
      this.domains.set(domain.id, domain);
    });
  }

  private initializeDefaultSites() {
    const defaultSites: Site[] = [
      {
        id: "all-sites",
        name: "All Sites",
        location: "Global",
        type: "global",
      },
      {
        id: "kumba",
        name: "Kumba Iron Ore",
        location: "South Africa",
        type: "iron-ore",
      },
      {
        id: "sishen",
        name: "Sishen Mine",
        location: "Northern Cape, South Africa",
        type: "iron-ore",
      },
      {
        id: "kolomela",
        name: "Kolomela Mine",
        location: "Northern Cape, South Africa",
        type: "iron-ore",
      },
      {
        id: "mogalakwena",
        name: "Mogalakwena PGMs",
        location: "Limpopo, South Africa",
        type: "platinum",
      },
      {
        id: "amandelbult",
        name: "Amandelbult Complex",
        location: "Limpopo, South Africa",
        type: "platinum",
      },
      {
        id: "minas-rio",
        name: "Minas-Rio",
        location: "Brazil",
        type: "iron-ore",
      },
      {
        id: "quellaveco",
        name: "Quellaveco",
        location: "Peru",
        type: "copper",
      },
      {
        id: "los-bronces",
        name: "Los Bronces",
        location: "Chile",
        type: "copper",
      },
      {
        id: "moranbah",
        name: "Moranbah North",
        location: "Queensland, Australia",
        type: "metallurgical-coal",
      },
      {
        id: "grosvenor",
        name: "Grosvenor",
        location: "Queensland, Australia",
        type: "metallurgical-coal",
      },
    ];

    defaultSites.forEach((site) => {
      this.sites.set(site.id, site);
    });
  }

  private initializeDefaultEndpoints() {
    const defaultEndpoints: Endpoint[] = [
      {
        id: "gpt-4-turbo",
        name: "GPT-4 Turbo",
        description: "OpenAI GPT-4 Turbo - Best for complex reasoning",
        type: "foundation",
        isDefault: true,
      },
      {
        id: "claude-3-5-sonnet",
        name: "Claude 3.5 Sonnet",
        description: "Anthropic Claude - Best for analysis and writing",
        type: "foundation",
        isDefault: false,
      },
      {
        id: "llama-3-1-70b",
        name: "Llama 3.1 70B",
        description: "Meta Llama - Open source, fast responses",
        type: "foundation",
        isDefault: false,
      },
      {
        id: "mining-ops-agent",
        name: "Mining Operations Agent",
        description: "Specialized agent for mining operations data",
        type: "agent",
        isDefault: false,
        domainId: "mining-ops",
      },
      {
        id: "geological-agent",
        name: "Geological Analysis Agent",
        description: "Agent for geological survey and exploration data",
        type: "agent",
        isDefault: false,
        domainId: "geological",
      },
      {
        id: "processing-agent",
        name: "Processing Plant Agent",
        description: "Agent for mineral processing optimization",
        type: "agent",
        isDefault: false,
        domainId: "processing",
      },
      {
        id: "sustainability-agent",
        name: "ESG & Sustainability Agent",
        description: "Agent for environmental and sustainability data",
        type: "agent",
        isDefault: false,
        domainId: "sustainability",
      },
      {
        id: "supply-chain-agent",
        name: "Supply Chain Agent",
        description: "Agent for logistics and procurement analytics",
        type: "agent",
        isDefault: false,
        domainId: "supply-chain",
      },
      {
        id: "finance-agent",
        name: "Finance Analytics Agent",
        description: "Agent for financial analysis and reporting",
        type: "agent",
        isDefault: false,
        domainId: "finance",
      },
    ];

    defaultEndpoints.forEach((endpoint) => {
      this.endpoints.set(endpoint.id, endpoint);
    });
  }

  async getConversations(): Promise<Conversation[]> {
    return Array.from(this.conversations.values());
  }

  async getConversation(id: string): Promise<Conversation | undefined> {
    return this.conversations.get(id);
  }

  async createConversation(endpointId: string, title: string, domainId?: string, siteId?: string): Promise<Conversation> {
    const id = randomUUID();
    const now = Date.now();
    const conversation: Conversation = {
      id,
      title,
      messages: [],
      endpointId,
      domainId,
      siteId,
      createdAt: now,
      updatedAt: now,
    };
    this.conversations.set(id, conversation);
    return conversation;
  }

  async addMessage(
    conversationId: string,
    message: Omit<Message, "id">
  ): Promise<Message> {
    const conversation = this.conversations.get(conversationId);
    if (!conversation) {
      throw new Error("Conversation not found");
    }

    const newMessage: Message = {
      ...message,
      id: randomUUID(),
    };

    conversation.messages.push(newMessage);
    conversation.updatedAt = Date.now();
    
    if (conversation.messages.length === 1 && message.role === "user") {
      conversation.title = message.content.slice(0, 50) + (message.content.length > 50 ? "..." : "");
    }

    return newMessage;
  }

  async updateConversation(
    id: string,
    updates: Partial<Conversation>
  ): Promise<Conversation | undefined> {
    const conversation = this.conversations.get(id);
    if (!conversation) return undefined;

    Object.assign(conversation, updates, { updatedAt: Date.now() });
    return conversation;
  }

  async deleteConversation(id: string): Promise<boolean> {
    return this.conversations.delete(id);
  }

  async getDomains(): Promise<Domain[]> {
    return Array.from(this.domains.values());
  }

  async getDomain(id: string): Promise<Domain | undefined> {
    return this.domains.get(id);
  }

  async createDomain(domain: InsertDomain): Promise<Domain> {
    let baseId = domain.name.toLowerCase().replace(/\s+/g, "-").replace(/[^a-z0-9-]/g, "");
    let id = baseId;
    let counter = 1;
    while (this.domains.has(id)) {
      id = `${baseId}-${counter}`;
      counter++;
    }
    const newDomain: Domain = { id, ...domain };
    this.domains.set(id, newDomain);
    return newDomain;
  }

  async updateDomain(id: string, updates: Partial<InsertDomain>): Promise<Domain | undefined> {
    const domain = this.domains.get(id);
    if (!domain) return undefined;
    const updated = { ...domain, ...updates };
    this.domains.set(id, updated);
    return updated;
  }

  async deleteDomain(id: string): Promise<boolean> {
    return this.domains.delete(id);
  }

  async getSites(): Promise<Site[]> {
    return Array.from(this.sites.values());
  }

  async getSite(id: string): Promise<Site | undefined> {
    return this.sites.get(id);
  }

  async getEndpoints(domainId?: string): Promise<Endpoint[]> {
    const allEndpoints = Array.from(this.endpoints.values());
    if (!domainId || domainId === "generic") {
      return allEndpoints.filter(e => !e.domainId || e.type === "foundation");
    }
    return allEndpoints.filter(e => !e.domainId || e.domainId === domainId || e.type === "foundation");
  }

  async getEndpoint(id: string): Promise<Endpoint | undefined> {
    return this.endpoints.get(id);
  }

  async createEndpoint(endpoint: InsertEndpoint): Promise<Endpoint> {
    let baseId = endpoint.name.toLowerCase().replace(/\s+/g, "-").replace(/[^a-z0-9-]/g, "");
    let id = baseId;
    let counter = 1;
    while (this.endpoints.has(id)) {
      id = `${baseId}-${counter}`;
      counter++;
    }
    const newEndpoint: Endpoint = { id, ...endpoint };
    this.endpoints.set(id, newEndpoint);
    return newEndpoint;
  }

  async updateEndpoint(id: string, updates: Partial<InsertEndpoint>): Promise<Endpoint | undefined> {
    const endpoint = this.endpoints.get(id);
    if (!endpoint) return undefined;
    const updated = { ...endpoint, ...updates };
    this.endpoints.set(id, updated);
    return updated;
  }

  async deleteEndpoint(id: string): Promise<boolean> {
    return this.endpoints.delete(id);
  }

  async getConfig(): Promise<Config> {
    return this.config;
  }

  async setConfig(config: Config): Promise<Config> {
    this.config = { ...this.config, ...config };
    return this.config;
  }
}

import { LakeBaseStorage, createLakeBaseConfig } from "./lakebase-storage";

let storageInstance: IStorage | null = null;

export async function initializeStorage(): Promise<IStorage> {
  if (storageInstance) return storageInstance;

  const lakebaseConfig = createLakeBaseConfig();
  
  if (lakebaseConfig) {
    console.log("LakeBase configuration found. Initializing LakeBase storage...");
    try {
      const lakebaseStorage = new LakeBaseStorage(lakebaseConfig);
      await lakebaseStorage.initialize();
      storageInstance = lakebaseStorage;
      console.log(`Connected to LakeBase: ${lakebaseConfig.catalog}.${lakebaseConfig.schema}`);
      return storageInstance;
    } catch (error) {
      console.error("Failed to initialize LakeBase storage, falling back to memory:", error);
    }
  }

  console.log("Using in-memory storage");
  storageInstance = new MemStorage();
  return storageInstance;
}

export function getStorage(): IStorage {
  if (!storageInstance) {
    storageInstance = new MemStorage();
  }
  return storageInstance;
}

File: backend/storage.py
from abc import ABC, abstractmethod
from typing import Optional
from uuid import uuid4
from .models import (
    Message, InsertMessage, Conversation, Domain, InsertDomain,
    Site, Endpoint, InsertEndpoint, Config, MessageRole, EndpointType
)
import time


class IStorage(ABC):
    @abstractmethod
    async def refresh_endpoints_from_databricks(self) -> list[Endpoint]:
        pass

    @abstractmethod
    async def get_conversations(self, user_email: Optional[str] = None) -> list[Conversation]:
        pass

    @abstractmethod
    async def get_conversation(self, id: str) -> Optional[Conversation]:
        pass

    @abstractmethod
    async def create_conversation(
        self, endpoint_id: str, title: str,
        domain_id: Optional[str] = None, site_id: Optional[str] = None,
        user_email: Optional[str] = None
    ) -> Conversation:
        pass

    @abstractmethod
    async def add_message(self, conversation_id: str, message: InsertMessage) -> Message:
        pass

    @abstractmethod
    async def update_conversation(self, id: str, updates: dict) -> Optional[Conversation]:
        pass

    @abstractmethod
    async def delete_conversation(self, id: str) -> bool:
        pass

    @abstractmethod
    async def get_domains(self) -> list[Domain]:
        pass

    @abstractmethod
    async def get_domain(self, id: str) -> Optional[Domain]:
        pass

    @abstractmethod
    async def create_domain(self, domain: InsertDomain) -> Domain:
        pass

    @abstractmethod
    async def update_domain(self, id: str, updates: dict) -> Optional[Domain]:
        pass

    @abstractmethod
    async def delete_domain(self, id: str) -> bool:
        pass

    @abstractmethod
    async def get_sites(self) -> list[Site]:
        pass

    @abstractmethod
    async def get_site(self, id: str) -> Optional[Site]:
        pass

    @abstractmethod
    async def get_endpoints(self, domain_id: Optional[str] = None) -> list[Endpoint]:
        pass

    @abstractmethod
    async def get_endpoint(self, id: str) -> Optional[Endpoint]:
        pass

    @abstractmethod
    async def create_endpoint(self, endpoint: InsertEndpoint) -> Endpoint:
        pass

    @abstractmethod
    async def update_endpoint(self, id: str, updates: dict) -> Optional[Endpoint]:
        pass

    @abstractmethod
    async def delete_endpoint(self, id: str) -> bool:
        pass

    @abstractmethod
    async def get_config(self) -> Config:
        pass

    @abstractmethod
    async def set_config(self, config: Config) -> Config:
        pass


class MemStorage(IStorage):
    def __init__(self):
        self.conversations: dict[str, Conversation] = {}
        self.domains: dict[str, Domain] = {}
        self.sites: dict[str, Site] = {}
        self.endpoints: dict[str, Endpoint] = {}
        self.config = Config()
        self._databricks_endpoints_loaded = False
        self._initialize_defaults()

    async def refresh_endpoints_from_databricks(self) -> list[Endpoint]:
        from .databricks_client import databricks_client
        
        if not databricks_client.is_configured():
            print("Databricks not configured, keeping default endpoints")
            return list(self.endpoints.values())
        
        try:
            db_endpoints = await databricks_client.list_serving_endpoints()
            
            if db_endpoints:
                self.endpoints.clear()
                for endpoint in db_endpoints:
                    self.endpoints[endpoint.id] = endpoint
                self._databricks_endpoints_loaded = True
                print(f"Loaded {len(db_endpoints)} endpoints from Databricks")
            else:
                print("No endpoints from Databricks, keeping defaults")
                
        except Exception as e:
            print(f"Error refreshing endpoints from Databricks: {e}")
            
        return list(self.endpoints.values())

    def _initialize_defaults(self):
        default_domains = [
            Domain(id="generic", name="General Assistant", description="General-purpose AI assistant for Anglo American", systemPrompt="You are a helpful AI assistant for Anglo American, a global mining company. Provide accurate, professional responses.", icon="Bot"),
            Domain(id="mining-ops", name="Mining Operations", description="Mining operations, production, and equipment management", systemPrompt="You are a mining operations specialist for Anglo American. Help with production optimization, equipment management, and operational efficiency.", icon="Pickaxe"),
            Domain(id="geological", name="Geological Services", description="Geological analysis, exploration, and resource estimation", systemPrompt="You are a geological services expert for Anglo American. Assist with geological analysis, exploration planning, and resource estimation.", icon="Mountain"),
            Domain(id="processing", name="Mineral Processing", description="Mineral processing and plant optimization", systemPrompt="You are a mineral processing specialist for Anglo American. Help optimize plant operations, throughput, and recovery rates.", icon="Factory"),
            Domain(id="sustainability", name="Sustainability & ESG", description="Environmental, social, and governance initiatives", systemPrompt="You are a sustainability and ESG advisor for Anglo American. Assist with environmental compliance, social responsibility, and governance reporting.", icon="Leaf"),
            Domain(id="supply-chain", name="Supply Chain", description="Supply chain, logistics, and procurement", systemPrompt="You are a supply chain specialist for Anglo American. Help with logistics optimization, procurement, and vendor management.", icon="Truck"),
            Domain(id="finance", name="Finance & Analytics", description="Financial analysis and business analytics", systemPrompt="You are a finance and analytics specialist for Anglo American. Assist with financial analysis, budgeting, and business intelligence.", icon="BarChart3"),
        ]
        for domain in default_domains:
            self.domains[domain.id] = domain

        default_sites = [
            Site(id="all-sites", name="All Sites", location="Global", type="Corporate"),
            Site(id="kumba", name="Kumba Iron Ore", location="South Africa", type="Iron Ore"),
            Site(id="sishen", name="Sishen Mine", location="Northern Cape, South Africa", type="Iron Ore"),
            Site(id="mogalakwena", name="Mogalakwena", location="Limpopo, South Africa", type="PGMs"),
            Site(id="unki", name="Unki Mine", location="Zimbabwe", type="PGMs"),
            Site(id="amandelbult", name="Amandelbult", location="Limpopo, South Africa", type="PGMs"),
            Site(id="quellaveco", name="Quellaveco", location="Peru", type="Copper"),
            Site(id="minas-rio", name="Minas-Rio", location="Brazil", type="Iron Ore"),
            Site(id="los-bronces", name="Los Bronces", location="Chile", type="Copper"),
            Site(id="moranbah", name="Moranbah", location="Queensland, Australia", type="Metallurgical Coal"),
            Site(id="sakatti", name="Sakatti", location="Finland", type="Copper-Nickel"),
            Site(id="woodsmith", name="Woodsmith", location="UK", type="Polyhalite"),
        ]
        for site in default_sites:
            self.sites[site.id] = site

        default_endpoints = [
            Endpoint(id="databricks-dbrx-instruct", name="DBRX Instruct", description="Databricks foundation model - fast and capable", type=EndpointType.foundation, isDefault=True),
            Endpoint(id="databricks-llama-3-70b", name="Llama 3 70B", description="Meta's Llama 3 70B model", type=EndpointType.foundation, isDefault=False),
            Endpoint(id="databricks-mixtral-8x7b", name="Mixtral 8x7B", description="Mistral AI mixture of experts", type=EndpointType.foundation, isDefault=False),
        ]
        for endpoint in default_endpoints:
            self.endpoints[endpoint.id] = endpoint

    async def get_conversations(self, user_email: Optional[str] = None) -> list[Conversation]:
        conversations = list(self.conversations.values())
        if user_email:
            conversations = [c for c in conversations if c.userEmail == user_email]
        return sorted(conversations, key=lambda c: c.updatedAt, reverse=True)

    async def get_conversation(self, id: str) -> Optional[Conversation]:
        return self.conversations.get(id)

    async def create_conversation(
        self, endpoint_id: str, title: str,
        domain_id: Optional[str] = None, site_id: Optional[str] = None,
        user_email: Optional[str] = None
    ) -> Conversation:
        conv_id = str(uuid4())
        now = int(time.time() * 1000)
        conversation = Conversation(
            id=conv_id,
            title=title,
            messages=[],
            endpointId=endpoint_id,
            domainId=domain_id,
            siteId=site_id,
            userEmail=user_email,
            createdAt=now,
            updatedAt=now
        )
        self.conversations[conv_id] = conversation
        return conversation

    async def add_message(self, conversation_id: str, message: InsertMessage) -> Message:
        conversation = self.conversations.get(conversation_id)
        if not conversation:
            raise ValueError("Conversation not found")
        
        msg_id = str(uuid4())
        new_message = Message(
            id=msg_id,
            role=message.role,
            content=message.content,
            timestamp=message.timestamp
        )
        conversation.messages.append(new_message)
        conversation.updatedAt = int(time.time() * 1000)
        return new_message

    async def update_conversation(self, id: str, updates: dict) -> Optional[Conversation]:
        conversation = self.conversations.get(id)
        if not conversation:
            return None
        
        for key, value in updates.items():
            if hasattr(conversation, key):
                setattr(conversation, key, value)
        conversation.updatedAt = int(time.time() * 1000)
        return conversation

    async def delete_conversation(self, id: str) -> bool:
        if id in self.conversations:
            del self.conversations[id]
            return True
        return False

    async def get_domains(self) -> list[Domain]:
        return list(self.domains.values())

    async def get_domain(self, id: str) -> Optional[Domain]:
        return self.domains.get(id)

    async def create_domain(self, domain: InsertDomain) -> Domain:
        base_id = domain.name.lower().replace(" ", "-")
        base_id = "".join(c for c in base_id if c.isalnum() or c == "-")
        domain_id = base_id
        counter = 1
        while domain_id in self.domains:
            domain_id = f"{base_id}-{counter}"
            counter += 1
        
        new_domain = Domain(id=domain_id, **domain.model_dump())
        self.domains[domain_id] = new_domain
        return new_domain

    async def update_domain(self, id: str, updates: dict) -> Optional[Domain]:
        domain = self.domains.get(id)
        if not domain:
            return None
        
        updated_data = domain.model_dump()
        updated_data.update(updates)
        updated_domain = Domain(**updated_data)
        self.domains[id] = updated_domain
        return updated_domain

    async def delete_domain(self, id: str) -> bool:
        if id in self.domains:
            del self.domains[id]
            return True
        return False

    async def get_sites(self) -> list[Site]:
        return list(self.sites.values())

    async def get_site(self, id: str) -> Optional[Site]:
        return self.sites.get(id)

    async def get_endpoints(self, domain_id: Optional[str] = None) -> list[Endpoint]:
        all_endpoints = list(self.endpoints.values())
        if not domain_id or domain_id == "generic":
            return [e for e in all_endpoints if not e.domainId or e.type == EndpointType.foundation]
        return [e for e in all_endpoints if not e.domainId or e.domainId == domain_id or e.type == EndpointType.foundation]

    async def get_endpoint(self, id: str) -> Optional[Endpoint]:
        return self.endpoints.get(id)

    async def create_endpoint(self, endpoint: InsertEndpoint) -> Endpoint:
        base_id = endpoint.name.lower().replace(" ", "-")
        base_id = "".join(c for c in base_id if c.isalnum() or c == "-")
        endpoint_id = base_id
        counter = 1
        while endpoint_id in self.endpoints:
            endpoint_id = f"{base_id}-{counter}"
            counter += 1
        
        new_endpoint = Endpoint(id=endpoint_id, **endpoint.model_dump())
        self.endpoints[endpoint_id] = new_endpoint
        return new_endpoint

    async def update_endpoint(self, id: str, updates: dict) -> Optional[Endpoint]:
        endpoint = self.endpoints.get(id)
        if not endpoint:
            return None
        
        updated_data = endpoint.model_dump()
        updated_data.update(updates)
        updated_endpoint = Endpoint(**updated_data)
        self.endpoints[id] = updated_endpoint
        return updated_endpoint

    async def delete_endpoint(self, id: str) -> bool:
        if id in self.endpoints:
            del self.endpoints[id]
            return True
        return False

    async def get_config(self) -> Config:
        return self.config

    async def set_config(self, config: Config) -> Config:
        self.config = config
        return self.config


storage_instance: Optional[IStorage] = None


async def initialize_storage() -> IStorage:
    global storage_instance
    
    from .lakebase_storage import create_lakebase_config, LakeBaseStorage
    
    lakebase_config = create_lakebase_config()
    if lakebase_config:
        try:
            lakebase_storage = LakeBaseStorage(lakebase_config)
            await lakebase_storage.initialize()
            storage_instance = lakebase_storage
            print("Using LakeBase storage (Databricks)")
        except Exception as e:
            print(f"Failed to initialize LakeBase storage: {e}")
            print("Falling back to in-memory storage")
            storage_instance = MemStorage()
    else:
        storage_instance = MemStorage()
        print("Using in-memory storage")
    
    await storage_instance.refresh_endpoints_from_databricks()
    
    return storage_instance


def get_storage() -> IStorage:
    global storage_instance
    if not storage_instance:
        storage_instance = MemStorage()
    return storage_instance
